modelci/finetuner/trainer.py

```diff
@@ -6,253 +6,85 @@
-            test_data_loader: torch.utils.data.DataLoader,
-            train_data_size: int = None,
-            test_data_size: int = None,
-            scheduler: torch.optim.lr_scheduler._LRScheduler = None,  # noqa
-            device: torch.device = None,
-    ):
-        """
-        Args:
-            net (nn.Module): PyTorch model.
-            train_data_loader (torch.utils.data.DataLoader): Data loader for training.
-            test_data_loader (torch.utils.data.DataLoader): Data loader for evaluating.
-            optimizer (torch.optim.optimizer.Optimizer): Optimizer.
-            train_data_size (int): Size of training data. Optional. Needed when using a sampler to split
-                training and validation data.
-            test_data_size (int): Size of evaluation data. Optional. Needed when using a sampler to split
-                training and validation data.
-        """
-        super().__init__(
-            net, optimizer=optimizer, train_data_loader=train_data_loader, test_data_loader=test_data_loader,
-            train_data_size=train_data_size, test_data_size=test_data_size, device=device,
-        )
-        self.scheduler = scheduler
-
-    def after_batch_finish(self):
-        if self.scheduler:
-            self.scheduler.step()
+    def __init__(self, model: pl.LightningModule, data_loader_kwargs: dict = None, trainer_kwargs: dict = None):
+        self.model = model
+        trainer_kwargs = trainer_kwargs or dict()
+        self.trainer_engine = pl.Trainer(**trainer_kwargs)
+        self._data_loader_kwargs = data_loader_kwargs or dict()
 
-    def train_one_batch(self, samples):
-        """Train one epoch.
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._event_pause = threading.Event()
+        self._task: Optional[Future] = None
 
-        Args:
-            samples: One batch of data
+    def start(self):
+        self._task = self._executor.submit(self.trainer_engine.fit, self.model, **self._data_loader_kwargs)
 
-        Returns:
-            Tuple of training loss and number of correctly predicted examples
-        """
-        inputs, labels = samples
-        inputs = inputs.to(self.device)
-        labels = labels.to(self.device)
+    def join(self, timeout=None):
+        if self._task:
+            self._task.result(timeout=timeout)
 
-        # forward + backward + optimize
-        outputs = self.net(inputs)
-        loss = self.criterion(outputs, labels)
-        loss.backward()
-        self.optimizer.step()
-        predicted_labels = torch.argmax(outputs, dim=1)
-        corrects = torch.sum(predicted_labels == labels)
+    def terminate(self):
+        if self._task:
+            # trigger pytorch lighting training graceful shutdown via a ^C
+            self._task.set_exception(KeyboardInterrupt())
 
-        return loss.item(), corrects.item()
+    def export_model(self):
+        return self.model.cpu()
```

Comment:
I'm wondering why should return  model object in CPU memory?

Reply:
When you save the model into the database, make sure all tensors reside in the memory.
If the model is in CUDA when saving, the user should remap the memory into other devices when loading. And this step may cause errors:
For instance, the model is trained on CUDA:3 and save. When a user does not have CUDA:3, he will receive a device error when loading.

Reply:
Thanks for your explanation